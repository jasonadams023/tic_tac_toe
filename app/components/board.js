import React from 'react';
import PropTypes from 'prop-types';

import Square from './square.js';

export default class Board extends React.Component {
  renderSquare(x, y) {
    const { store } = this.context;
    const state = store.getState();
    const history = state.history;
    const current = history[state.stepNumber];
    const squares = current.squares;

    return  (
                <Square
                    key={ x + '-' + y }
                    value={ squares[x][y] }
                    onClick={ () => this.handleClick(x, y) }
                />
            );
  }

  handleClick(x, y) {
    const { store } = this.context;

    store.dispatch({ type: "PLAYER_MOVE", x: x, y: y });
  }

  render() {
    const { store } = this.context;
    const state = store.getState();
    const history = state.history;
    const current = history[state.stepNumber];
    const squares = current.squares;

    return (
        <div className="game-board" >
            {
                squares.map((row, xIndex) => {
                    return <div key={ xIndex } className="board-row">
                        {
                            row.map((space, yIndex) => {
                                return this.renderSquare(xIndex, yIndex)
                            })
                        }
                    </div>
                })
            }
        </div>
    );
  }
}

Board.contextTypes = {
    store: PropTypes.object
}