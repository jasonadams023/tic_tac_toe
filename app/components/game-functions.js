export function calculateWinner(squares) {
    ////check rows
    for (let x = 0; x < squares.length; x++) {
        let winner = squares[x][0];
        for (let y = 0; y < squares[x].length - 1; y++) {
            if (winner !== squares[x][y + 1]) {
                winner = null;
            }
        }

        if (winner !== null) {
            return winner;
        }
    }


    ////check columns
    for (let y = 0; y < squares[0].length; y++) {
        let winner = squares[0][y];
        for (let x = 0; x < squares.length - 1; x++) {
            if (winner !== squares[x + 1][y]) {
                winner = null;
            }
        }

        if (winner !== null) {
            return winner;
        }
    }

    ////check diagonals
    let winner = squares[0][0];
    let x = 1;
    let y = 1;
    for (; x < squares.length && y < squares[0].length; x++, y++) {
        if (winner !== squares[x][y]) {
            winner = null;
        }
    }

    if (winner !== null) {
        return winner;
    }

    x = squares.length - 1;
    y = 0;
    winner = squares[x][y];
    for (; x >= 0 && y < squares[x].length; x--, y++) {
        if (winner !== squares[x][y]) {
            winner = null;
        }
    }

    if (winner !== null) {
        return winner;
    }


    return null;
}

export function calculateTie(squares) {
    for (let x = 0; x < squares.length; x++) {
        for (let y = 0; y < squares[x].length; y++) {
            if (squares[x][y] === null) {
                return false;
            }
        }
    }

    return true;
}

export function generateBoard(size) {
    let rows = [];
    let board = [];

    for (let i = 0; i < size; i++) {
        rows.push(Array(size).fill(null));
        board.push(rows[i]);
    }

    return board;
}