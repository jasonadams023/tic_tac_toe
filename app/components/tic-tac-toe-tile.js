import React from 'react';
import PropTypes from 'prop-types';

import Square from './square';

export default class TicTacToeTile extends React.Component {
    render () {
        const { store } = this.context;
        const state = store.getState();
        const x = this.props.coordinates.x;
        const y = this.props.coordinates.y;
        const history = state.history;
        const current = history[state.stepNumber];
        const squares = current.squares;
        const value = squares[x][y];

        const handleClick = (x, y) => { store.dispatch({ type: "PLAYER_MOVE", x: x, y: y }) };

        return  (
            <Square
                value={ value }
                onClick={ () => handleClick(x, y) }
            />
        );
    }
}

TicTacToeTile.contextTypes = {
    store: PropTypes.object
}