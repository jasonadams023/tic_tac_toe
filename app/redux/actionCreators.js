export const changeStep = (step) => {
    return { type: "CHANGE_STEP", stepNumber: step };
};

export const playerMove = (x, y) => {
    return { type: "PLAYER_MOVE", x: x, y: y };
};