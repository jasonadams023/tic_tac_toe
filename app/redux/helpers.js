import { generateBoard } from '../components/game-functions';

export function stateUndefined(state) {
    return typeof state === 'undefined' ? true : false;
}

export function defaultState() {
    return {
        history: [{squares: generateBoard(3)}],
        stepNumber: 0
    };
}