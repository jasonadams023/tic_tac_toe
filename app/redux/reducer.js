/*
State:
    history: [ {squares: boardState} ],
    stepNumber: int
*/

import { calculateWinner, generateBoard, nextPlayer } from '../components/game-functions.js';

export function reducer (state, action) {
    if (typeof state === 'undefined') {
        state = {
            history: [ {squares: generateBoard(3)} ],
            stepNumber: 0
        };
    }

    if (action.type === "PLAYER_MOVE") {
        const oldHistory = state.history.slice();
        const currentMove = oldHistory[oldHistory.length - 1];
        let squares = [];
        const x = action.x;
        const y = action.y;

        for (let i = 0; i < currentMove.squares.length; i++) {
            squares.push(currentMove.squares[i].slice());
        }

        if (calculateWinner(squares) || squares[x][y]) {
            return state;
        }

        squares[x][y] = nextPlayer(state.stepNumber);

        const newHistory = oldHistory.concat([{ squares: squares }]);
        let nextStep = state.stepNumber + 1;

        return Object.assign({}, state, {history: newHistory, stepNumber: nextStep});
    }
    if (action.type === "CHANGE_STEP") {
        return Object.assign({}, state, {stepNumber: action.stepNumber});
    }

    return state;
}