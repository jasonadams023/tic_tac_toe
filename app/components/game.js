import React from 'react';
import { Provider } from 'react-redux';

import Board from './board.js';
import { calculateWinner, calculateBoardFull, generateBoard, nextPlayer } from './game-functions.js';
import { store } from '../redux/store';
import SubscribedComponent from '../redux/subscribedComponent';
import GameInfo from './game-info/game-info';

export default class Game extends SubscribedComponent {
  constructor() {
    super();
  }

  handleClick(x, y) {
    store.dispatch({ type: "PLAYER_MOVE", x: x, y: y });
  }

  jumpTo(step) {
    store.dispatch({ type: "CHANGE_STEP", stepNumber: step });
  }

  status(squares) {
      const winner = calculateWinner(squares);

      if (winner) {
          return 'Winner is: ' + winner;
      } else if (calculateBoardFull(squares)) {
          return "Cat's Game";
      }else {
          return 'Next player: ' + nextPlayer(store.getState().stepNumber);
      }
  }

  getMoves(history) {
    return history.map((step, move) => {
           const description = move ? 'Move #' + move : 'Game Start';

           return (
               <li key={ move }>
                   <a href='#' className="moves" onClick={ () => this.jumpTo(move) }> { description } </a>
               </li>
           );
       });
  }

  render() {
    const state = store.getState();
    const history = state.history;
    const current = history[state.stepNumber];
    const moves = this.getMoves(history);

    return (
      <div className="game">
        <div className="game-board">
          <Board squares={ current.squares }
                 onClick={ (x, y) => this.handleClick(x, y) }
          />
        </div>
        <div className="game-info">
          <div id="status">{ this.status(current.squares) }</div>
          <ol>{ moves }</ol>
        </div>
      </div>
    );
  }
}