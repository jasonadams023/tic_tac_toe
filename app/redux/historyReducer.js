import { calculateWinner, generateBoard, nextPlayer } from '../components/game-functions.js';
import { defaultState } from './helpers';

export function historyReducer(state = defaultState(), action) {
    if (action.type === "PLAYER_MOVE") {
            const oldHistory = state.history.slice(0, state.stepNumber + 1);
            const currentMove = oldHistory[oldHistory.length - 1];
            let squares = [];
            const x = action.x;
            const y = action.y;

            for (let i = 0; i < currentMove.squares.length; i++) {
                squares.push(currentMove.squares[i].slice());
            }

            if (calculateWinner(squares) || squares[x][y]) {
                return state;
            }

            squares[x][y] = nextPlayer(state.stepNumber);

            const newHistory = oldHistory.concat([{ squares: squares }]);
            let nextStep = state.stepNumber + 1;

            return Object.assign({}, state, {history: newHistory, stepNumber: nextStep});
        }

    return state;
}