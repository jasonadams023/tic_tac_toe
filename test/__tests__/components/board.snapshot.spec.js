import React from 'react';
import Board from '../../../app/components/board';
import { generateBoard } from '../../../app/components/game-functions';
import { Provider } from 'react-redux';
import { store } from '../../../app/redux/store';


describe('Board', () => {
    let squares;
    let x = 'X';
    let o = 'O';

    beforeEach(() => {
        squares = generateBoard(3);
    })

    it('displays 3x3 set of squares', () => {
        const wrapper = render (
            < Provider store={ store } >
                < Board squares={ squares } />
            </Provider>
        );

        expect(wrapper).toMatchSnapshot();
    });

    it('passes values of squares to squares', () => {
        let action = { type: "PLAYER_MOVE", x: 0, y: 0 };
        store.dispatch(action);
        action.y = 1;
        store.dispatch(action);
        action.y = 2;
        store.dispatch(action);
        action.y = 1;
        action.x = 1;
        store.dispatch(action);
        action.y = 0;
        store.dispatch(action);
        action.x = 2;
        store.dispatch(action);
        action.y = 2;
        action.x = 1;
        store.dispatch(action);
        action.x = 2;
        action.y = 2;
        store.dispatch(action);
        action.y = 1;
        store.dispatch(action);

        const wrapper = render (
            < Provider store={ store } >
                < Board />
            </Provider>
        );

        expect(wrapper).toMatchSnapshot();
    })
});