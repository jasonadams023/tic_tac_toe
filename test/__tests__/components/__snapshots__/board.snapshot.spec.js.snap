// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board displays 3x3 set of squares 1`] = `
<div
  class="game-board"
>
  <div
    class="board-row"
  >
    <button
      class="square"
    />
    <button
      class="square"
    />
    <button
      class="square"
    />
  </div>
  <div
    class="board-row"
  >
    <button
      class="square"
    />
    <button
      class="square"
    />
    <button
      class="square"
    />
  </div>
  <div
    class="board-row"
  >
    <button
      class="square"
    />
    <button
      class="square"
    />
    <button
      class="square"
    />
  </div>
</div>
`;

exports[`Board passes values of squares to squares 1`] = `
<div
  class="game-board"
>
  <div
    class="board-row"
  >
    <button
      class="square"
    >
      X
    </button>
    <button
      class="square"
    >
      O
    </button>
    <button
      class="square"
    >
      X
    </button>
  </div>
  <div
    class="board-row"
  >
    <button
      class="square"
    >
      X
    </button>
    <button
      class="square"
    >
      O
    </button>
    <button
      class="square"
    >
      X
    </button>
  </div>
  <div
    class="board-row"
  >
    <button
      class="square"
    >
      O
    </button>
    <button
      class="square"
    >
      X
    </button>
    <button
      class="square"
    >
      O
    </button>
  </div>
</div>
`;
